src/AbstractBlocklockReceiver.sol

```diff
@@ -1,20 +1,196 @@
+        );
+    }
+
+    /// @notice Requests a blocklock with a subscription and returns the request ID.
+    /// @dev This function calls the `requestBlocklock` function from the `blocklock` contract, passing the required parameters such as
+    ///      `callbackGasLimit`, `subscriptionId`, `blockHeight`, and `ciphertext`.
+    /// @param callbackGasLimit The gas limit for the callback function to be executed after the blocklock request.
+    /// @param blockHeight The block height for which the blocklock request is made.
+    /// @param ciphertext The ciphertext to be used in the blocklock request.
+    /// @return requestId The unique identifier for the blocklock request.
+    /// @notice This function internally calls the `blocklock.requestBlocklock` function.
+    function _requestBlocklockWithSubscription(
+        uint32 callbackGasLimit,
+        uint256 blockHeight,
+        TypesLib.Ciphertext calldata ciphertext
+    ) internal returns (uint256 requestId) {
+        return blocklock.requestBlocklock(callbackGasLimit, subscriptionId, blockHeight, ciphertext);
+    }
+
+    /// @notice Decrypts the provided ciphertext using the specified decryption key.
+    /// @dev This function calls the `decrypt` function from the `blocklock` contract to perform the decryption operation.
+    ///      It requires that the `blocklock` contract implements decryption logic using the provided ciphertext and decryption key.
+    /// @param ciphertext The ciphertext that needs to be decrypted.
+    /// @param decryptionKey The decryption key to be used for decrypting the ciphertext.
+    /// @return The decrypted plaintext as a `bytes` array.
+    /// @notice This function internally calls the `blocklock.decrypt` function to perform the decryption.
+    function _decrypt(TypesLib.Ciphertext memory ciphertext, bytes calldata decryptionKey)
+        internal
+        view
+        returns (bytes memory)
+    {
+        return blocklock.decrypt(ciphertext, decryptionKey);
+    }
+
+    /// @notice Handles the reception of a blocklock with the provided decryption key.
+    /// @dev This function is meant to be overridden in derived contracts to define the specific logic
+    ///      for processing a blocklock upon receipt of a decryption key.
+    /// @param requestID The unique identifier of the blocklock request.
+    /// @param decryptionKey The decryption key that corresponds to the ciphertext in the blocklock request.
+    /// @notice This function does not implement any functionality itself but serves as a placeholder for derived contracts
+    ///         to implement their specific logic when a blocklock is received.
+    /// @dev This function is marked as `internal` and `virtual`, meaning it can be overridden in a derived contract.
+    function _onBlocklockReceived(uint256 requestID, bytes calldata decryptionKey) internal virtual;
```

Comment:
feels weirdly to prefix functions that you have to override with `_` - normally I feel like if I'm touching a `_function` of somebody else's code I'm doing something wrong

Reply:
The usual convention is to use an underscore prefix for non-external functions and state variables but I can update depending on preference @CluEleSsUK 

https://docs.soliditylang.org/en/latest/style-guide.html#underscore-prefix-for-non-external-functions-and-variables

Reply:
ahh - haven't seen that style guide. In that case, let's stick with it!